### PR TITLE
Annotate each C++ function with a macro

### PR DESCRIPTION
**Pull request by dcrc2**
As explained in this comment at the top of knossos.h:

```
Each C++ function is annotated with one of the following macros:
- KS_DEF: a definition which has been generated by ksc
- KS_FUNCTION: a function used only from within ksc-generated code
- KS_INLINE_FUNCTION: an inline KS_FUNCTION
- KS_INTERFACE: a function which may be called from outside the generated code (e.g. to convert to/from ks types)
```

The aim is to allow the generated C++ files to be used as CUDA files without modifying ksc. This can be achieved by `#define KS_DEF __device__`, and similarly for the other macros. The CUDA support is not added by this PR: I've just added the macros on their own to confirm that there are no breaking changes.

### CODE REVIEW

File: src/runtime/knossos.h [hunk 1]
```diff
@@ -89,54 +102,54 @@ namespace ks {
...
+	template<> struct get_helper_t<0> { template<typename Tuple> static KS_INTERFACE auto const& get(Tuple const& t) { return t.t0; } };
+	template<> struct get_helper_t<1> { template<typename Tuple> static KS_INTERFACE auto const& get(Tuple const& t) { return t.t1; } };
+	template<> struct get_helper_t<2> { template<typename Tuple> static KS_INTERFACE auto const& get(Tuple const& t) { return t.t2; } };
+	template<> struct get_helper_t<3> { template<typename Tuple> static KS_INTERFACE auto const& get(Tuple const& t) { return t.t3; } };
 
-	template<typename T> T& const_cast_deduce_type(T const& t) { return const_cast<T&>(t); }
+	template<typename T> KS_INTERFACE T& const_cast_deduce_type(T const& t) { return const_cast<T&>(t); }
 
-	template<size_t I, typename ...Ts> auto& get(Tuple<Ts...> & t) {
+	template<size_t I, typename ...Ts> KS_INTERFACE auto& get(Tuple<Ts...> & t) {
```
Comment (awf) [inline src/runtime/knossos.h:130]: I think we could/should have the convention that KS_INTERFACE etc should begin a new line

**Comment by awf** [comment]
I find KS_INLINE_FUNCTION a bit odd.  Do the others not have INLINE versions because we know they will never be needed, or because they are not currently needed?  

Perhaps the opening comment could also show the CUDA expansion for each macro?

**Comment by awf** [comment]
Would a separate KS_INLINE work?

**Comment by awf** [comment]
> This can be achieved by #define KS_DEF __device__, and similarly for the other macros. 

Maybe add some of this discussion to the opening comment

**Comment by dcrc2** [comment]
> I find KS_INLINE_FUNCTION a bit odd. Do the others not have INLINE versions because we know they will never be needed, or because they are not currently needed?

Not currently needed - the `KS_INTERFACE` functions are all either templates or class member functions.

I can replace with `inline KS_FUNCTION` if that looks better. My only worry was having to decide between `inline KS_FUNCTION` and `KS_FUNCTION inline`: in fact both of these work for the CUDA support, but I wasn't sure whether if we wanted to play around with adding other attributes, they might be more picky about whether they appear before or after the `inline` keyword.

**Comment by dcrc2** [comment]
> I can replace with `inline KS_FUNCTION` if that looks better.

I've pushed a change to do it this way.

File: src/runtime/knossos.h (hunk 1 above)
Comment (dcrc2) [reply src/runtime/knossos.h]: OK done

**Comment by dcrc2** [comment]
> > This can be achieved by `#define KS_DEF __device__`, and similarly for the other macros.
> 
> Maybe add some of this discussion to the opening comment

OK I've pulled in another commit from #976, which defines these macros for CUDA and non-CUDA code. These definitions come immediately after the comment so hopefully this makes the purpose more clear.

File: src/runtime/knossos.h [hunk 2]
```diff
@@ -33,7 +33,29 @@
...
+- KS_INTERFACE: a function which may be called from outside the generated code (e.g. to convert to/from ks types)
+*/
+
+#ifdef KS_CUDA
+
+#define KS_DEF __device__
+#define KS_FUNCTION __device__
+#define KS_INTERFACE __host__ __device__
+
+#else
```
Comment (awf) [inline src/runtime/knossos.h:49]: Yes, much clearer to me.

**Review by awf — APPROVED** [review]
Great, thanks.

**Pull request merged**